geode-core/src/main/java/org/apache/geode/management/internal/operation/OperationHistoryManager.java

```diff
@@ -36,30 +36,55 @@
  */
 @Experimental
 public class OperationHistoryManager {
-  private final ConcurrentMap<String, OperationInstance> inProgressHistory;
-  private final Map<String, OperationInstance> completedHistory;
+  private final ConcurrentMap<String, OperationInstance> history;
+  private final long keepCompletedMillis;
 
+  /**
+   * set a default retention policy to keep results for 2 hours after completion
+   */
   public OperationHistoryManager() {
-    this(100);
+    this(2, TimeUnit.HOURS);
   }
 
-  @SuppressWarnings("unchecked")
-  public OperationHistoryManager(int historySize) {
-    inProgressHistory = new ConcurrentHashMap<>();
-    completedHistory = Collections.synchronizedMap(new LRUMap(historySize));
+  /**
+   * set a custom retention policy to keep results for X amount of time after completion
+   */
+  public OperationHistoryManager(long keepCompleted, TimeUnit timeUnit) {
+    history = new ConcurrentHashMap<>();
+    keepCompletedMillis = timeUnit.toMillis(keepCompleted);
   }
 
   /**
-   * check both maps for the specified key
+   * look up the specified key
    */
   @SuppressWarnings("unchecked")
   <A extends ClusterManagementOperation<V>, V extends JsonSerializable> OperationInstance<A, V> getOperationInstance(
       String opId) {
-    OperationInstance operationInstance = inProgressHistory.get(opId);
-    if (operationInstance == null) {
-      operationInstance = completedHistory.get(opId);
+    expireHistory();
+    return history.get(opId);
+  }
+
+  private void expireHistory() {
+    final long purgeThreshold = System.currentTimeMillis() - keepCompletedMillis;
+    Set<String> expiredKeys = history.entrySet().stream()
+        .filter(e -> e.getValue().getFutureOperationEnded().isDone()
+            && getOperationEnded(e.getValue()).getTime() < purgeThreshold)
+        .map(e -> e.getKey()).collect(Collectors.toSet());
+    expiredKeys.forEach(history::remove);
+  }
+
+  /**
+   * @return null if unavailable
+   */
+  private static Date getOperationEnded(OperationInstance<?, ?> operationInstance) {
```

Comment:
how about making this "boolean hasExpired(operationInstance, purgeThreshold)"
Then you can do the test against purgeThreshold here and if you are not able to get a Date
because of the exceptions return either true or false instead of null.
Returning null will cause the caller of getOperationEnded to throw NPE since it always calls getTime() on the result.
It seems like it would be okay to return true and remove items whose get() method is throwing.

Reply:
In the usage it never would actually returns null, because isDone is checked first.  However, a `hasExpired` test is a little cleaner, so went ahead and made that change